Fix client metrics recording on round trip error

alimikegami opened the pull request:
This pull request fixes client metrics recording on round trip error

Related issues: #6940 

Does this approach correctly address the issue with round trip errors?
Should I add specific tests to verify metric recording (`http.client.duration`) during round trip errors?

dmathieu commented:
Duplicating code like isn't really maintainable long-term. We should be able to have the metrics recorded, with no code duplication. Possibly, by exporting the metric into its own method.

Also, this needs to be tested.

alimikegami commented:
Thank you for the review @dmathieu 

What do you think of my latest commit? I’ve added a method to record metrics and removed most of the unnecessary code duplication. Thank you.

dmathieu commented on instrumentation/net/http/otelhttp/transport.go:
```diff
@@ -141,15 +151,20 @@ func (t *Transport) RoundTrip(r *http.Request) (*http.Response, error) {
+		metricOpts := t.semconv.MetricOptions(metricAttributes)
+
+		// set transport record metrics
+		t.recordMetrics(ctx, requestStartTime, metricData, metricOpts)
```
How about calling this before the error check, in a `defer` statement?

dmathieu reviewed (commented):
Could you add a test?

alimikegami commented:
Thank you for the review @dmathieu 
I've added the test and refactored the metric recording to use defer, removing the remaining code duplication.

Please check my latest commit. Thank you!

dmathieu commented:
This should close https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6940

alimikegami commented:
> @alimikegami the CI has been failing since your last commit. Could you fix it?

I'll fix it as soon as possible!

alimikegami commented:
<img width="1100" alt="image" src="https://github.com/user-attachments/assets/ab3b6035-9c8a-46a2-9312-704504cd40ef" />

<br>
<br>

Hi @dmathieu,

I've run the tests on my local machine, but I’m not seeing the `http.client.request.body.size` metric that appears in the CI test results. Instead, I'm seeing `http.client.request.size`. The same also applies to the duration metric.

Here's the test output from my local environment:
<br>
<img width="1095" alt="image" src="https://github.com/user-attachments/assets/782b1848-4e57-4085-8787-c9ce771bd8ff" />

<br>

Do you know what might be causing the difference in metric names between the CI environment and my local machine?

dmathieu commented:
This error would hint on the semconv migration. Is your commit based on a rather old commit in this repo?
The proper metric is the one this CI sees (see https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md).

alimikegami commented:
Thank you for the explanation, @dmathieu. It cleared up my confusion. Please kindly review my test and implementation! I have fixed the failing tests. Thank you!

dmathieu commented on CHANGELOG.md:
```diff
@@ -75,6 +75,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - `go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo`.
   - `go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace`.
   - `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`.
+- Record `http.client.request.body.size` and `http.client.request.duration` metrics on transport round trip errors in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` (#6940)
```
```suggestion
- Record all non-failure metrics on transport round trip errors in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` (#7146)
```

dmathieu commented on instrumentation/net/http/otelhttp/transport.go:
```diff
@@ -141,36 +169,26 @@ func (t *Transport) RoundTrip(r *http.Request) (*http.Response, error) {
+	return res, nil
+}
 
+func (t *Transport) recordMetrics(ctx context.Context, requestStartTime time.Time, metricData semconv.MetricData, metricOpts map[string]semconv.MetricOpts) {
```
This method is only called in the defer. Does it need to be a method?

alimikegami replied:
It might not be necessary at the moment, I'll remove the method. Thank you.

alimikegami commented:
I've committed the CHANGELOG update and removed the `recordMetrics` method. Please kindly review the changes. Thank you!

dmathieu commented on CHANGELOG.md:
```diff
@@ -85,6 +85,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - `go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo`.
   - `go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace`.
   - `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`.
+- Record all non-failure metrics on transport round trip errors in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` (#7146)
```
With yesterday's release, this is now in the released section. It needs to be moved up to unreleased.

alimikegami replied:
I've moved the entry to the unreleased section. Please kindly check my latest commit. Thank you!

alimikegami commented on instrumentation/net/http/otelhttp/transport.go:
```diff
@@ -129,6 +129,38 @@ func (t *Transport) RoundTrip(r *http.Request) (*http.Response, error) {
+	metricAttributes := semconv.MetricAttributes{
+		Req:                  r,
+		AdditionalAttributes: append(labeler.Get(), t.metricAttributesFromRequest(r)...),
+	}
```
I don’t believe there was any particular reason it wasn’t in the deferred function. I’ve moved it to the deferred function in my latest commit. Thank you for pointing it out!

alimikegami commented:
I've refactored the code according to your suggestions, @pellared. Please have a look at my latest commit. Thank you!

pellared reviewed (dismissed):
@dmathieu, can you take another look please?

dmathieu commented:
The latest changes appear to break the tests.

alimikegami commented:
I've fixed the failing tests, please kindly check my latest commit. Thank you!
@dmathieu @pellared
